### PR TITLE
Support accessing a model's metadata in C#

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**:Support accessing a model's metadata in C#

**Motivation and Context**
Resolve https://github.com/microsoft/onnxruntime/issues/4568

### CODE REVIEW

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs [hunk 1]
```diff
@@ -1563,6 +1563,36 @@ private void TestModelSequenceOfTensors()
             }
         }
 
+        [Fact]
+        private void TestModelMetadata()
+        {
+
+            string modelPath = Path.Combine(Directory.GetCurrentDirectory(), "model_with_valid_ort_config_json.onnx");
```
Comment (Craigacp) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1570]: I should port this test over to Java, I didn't create a model with custom fields to test out the Java metadata support.

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 2]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
...
+            IntPtr graphNameHandle = IntPtr.Zero;
+            IntPtr domainHandle = IntPtr.Zero;
+            IntPtr descriptionHandle = IntPtr.Zero;
+
+            var allocator = OrtAllocator.DefaultInstance;
+            try
+            {
+                // Get the native ModelMetadata instance associated with the InferenceSession
+
+                NativeApiStatus.VerifySuccess(NativeMethods.OrtSessionGetModelMetadata(session.Handle, out modelMetadataHandle));
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1024]: This should go before try

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 3]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
...
+        private string _domain;
+        private string _description;
+        private long _version;
+        private Dictionary<string, string> _customMetadataMap = new Dictionary<string, string>();
+
+        internal ModelMetadata(InferenceSession session)
+        {
+            IntPtr modelMetadataHandle = IntPtr.Zero;
+
+            IntPtr producerNameHandle = IntPtr.Zero;
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1014]: >IntPtr producerNameHandle = IntPtr.Zero; [](start = 12, length = 40)

This should go right before OrtModelMetadataGetProducerName() call

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 4]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
...
+            IntPtr graphNameHandle = IntPtr.Zero;
+            IntPtr domainHandle = IntPtr.Zero;
+            IntPtr descriptionHandle = IntPtr.Zero;
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1017]: Ditto

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 5]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
...
+                            // Put the key/value pair into the dictionary
+                            _customMetadataMap[key] = value;
+
+                        }
+
+                        finally
+                        {
+                            // Free the memory associated with this single key value pair
+                            allocator.FreeMemory(keyHandle);
+                            allocator.FreeMemory(valueHandle);
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1097]: Why not wrap it in using() with OrtMemoryAllocation()? Simlper code

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 6]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
...
+
+            }
+
+            finally
+            {
+                // producerHandle, graphNameHandle,  domainHandle, descriptionHandle will be released when the corresponding OrtMemoryAllocation instance above gets disposed
+
+                // Free ModelMetadata handle
+                if (modelMetadataHandle != IntPtr.Zero)
+                {
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1111]: Remove the check for zero. Try would not be entered if we do not successfully acquire modelMetadataHandle

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 7]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
...
+
+
+                // Process CustomMetadata Map
+
+                IntPtr customMetadataMapKeysHandle = IntPtr.Zero;
+                long numKeys;
+                NativeApiStatus.VerifySuccess(NativeMethods.OrtModelMetadataGetCustomMetadataMapKeys(modelMetadataHandle, allocator.Pointer, out customMetadataMapKeysHandle, out numKeys));
+
+                // We have received an array of null terminated C strings which are the keys that we can use to lookup the custom metadata map
+                // The OrtAllocator will finally free the customMetadataMapKeysHandle
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1071]: I will need clarification on this one.

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 8]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
     }
 
 
-    internal class ModelMetadata
+    public class ModelMetadata
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1001]: >   public class ModelMetadata [](start = 1, length = 29)

You have chosen to implement the class in a way that queries everything. Another way to imlpement it is to query things on demand, only when needed. This is the way OrtMemoryInfo is implemented so the user pays only for what is needed.
Have you considered doing it this way and not to cache anything in managed memory? What is the use pattern here?

Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1001]: I thought about the "pay for what you want" approach but as synced offline - I feel that it doesn't provide too much value as folks are most likely going to require most of the metadata  anyway and we would have to cache the model metadata native handle for the lifetime of the instance and dispose it at the end - I think this approach works for this kind of a class.

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs (hunk 2 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1024]: done

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs (hunk 7 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1071]: Re-wrote it as we discussed offline

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs (hunk 1 above)
Comment (hariharans29) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests/InferenceTest.cs:1570]: you can use the test model I have here.

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs (hunk 5 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1097]: Yes, have that now - thank you!

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs [hunk 9]
```diff
@@ -993,9 +998,171 @@ public bool IsTensor
...
+        private string _description;
+        private long _version;
+        private Dictionary<string, string> _customMetadataMap = new Dictionary<string, string>();
+
+        internal ModelMetadata(InferenceSession session)
+        {
+            IntPtr modelMetadataHandle = IntPtr.Zero;
+
+            IntPtr producerNameHandle = IntPtr.Zero;
+            IntPtr graphNameHandle = IntPtr.Zero;
```
[the remark this answers is not in the document]
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1015]: done

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs (hunk 4 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1017]: done

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs (hunk 3 above)
Comment (hariharans29) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs:1014]: done

**Pull request merged**